src/stravalib/protocol.py

```diff
@@ -89,6 +102,187 @@ def __init__(
         )
 
+    # (pylance?) mypy is flagging this method. Not sure why.
+
+    def _check_credentials(self) -> tuple[int, str] | None:
+        """Gets Strava client_id and secret credentials from user's environment.
+
+        If the user environment is populated with both values and
+        client_id is a proper int, it returns a tuple with both values.
+        Otherwise it returns None.
+
+        Returns
+        -------
+        Tuple or None
+            Tuple with the Client id and secret if the values exist in the users
+            environment, otherwise None.
+
+        """
+
+        client_id_str = os.environ.get("STRAVA_CLIENT_ID")
+        client_secret = os.environ.get("STRAVA_CLIENT_SECRET")
+        # Make sure client_id exists and can be cast to int
+        if client_id_str:
+            try:
+                # Make sure client_id is a valid int
+                client_id: int = int(client_id_str)
+            except ValueError:
+                logging.error("STRAVA_CLIENT_ID must be a valid integer.")
+                return None
+
+        if client_id and client_secret:
+            return client_id, client_secret
+        else:
+            logging.warning(
+                "STRAVA_CLIENT_ID and STRAVA_CLIENT_SECRET not found in your "
+                " environment. Please refresh your access_token manually."
+                " Or add STRAVA_CLIENT_ID and STRAVA_CLIENT_SECRET to your environment."
+            )
+            return None
+
+    def _request(
+        self,
+        url: str,
+        params: dict[str, Any] | None = None,
+        files: dict[str, SupportsRead[str | bytes]] | None = None,
+        method: RequestMethod = "GET",
+        check_for_errors: bool = True,
+    ) -> Any:
+        """Perform the underlying request, returning the parsed JSON results.
+
+        Before running, this method will check to make sure that your
+        access_token is valid. If it isn't, it will refresh it and then make
+        the requested Strava API call.
+
+        Parameters
+        ----------
+        url : str
+            The request URL.
+        params : Dict[str,Any]
+            Request parameters
+        files : Dict[str,file]
+            Dictionary of file name to file-like objects.
+        method : str
+            The request method (GET/POST/etc.)
+        check_for_errors : bool
+            Whether to raise an error or not.
+
+        Returns
+        -------
+        Dict[str, Any]
+            The parsed JSON response.
+        """
+
+        # If the request is a token request, skip, otherwise
+        # refresh token
+        # This feels sloppy but if check credentials is None then that means
+        # users environment isn't setup
+        credentials = self._check_credentials()
+
+        # Only refresh token if we know the users' environment is setup
+        if "/oauth/token" not in url and credentials:
+            self.refresh_expired_token(credentials[0], credentials[1])
```

Comment:
This is what i don't like. I wanted to unpack a tuple with client_id, client_secret. It would make the code more expressive. The alternative would be that the tuple contains a third value that is a bool for valid credentials? that would make that method a bit more complex but would make the code here more expressive.

Reply:
If you move the check and extraction of credentials to `APIV3`s initializer, this should be no problem anymore.

Reply:
it's much better! thank you for catching that. it just gets assigned to self now which makes everything simpler.